File: sliding_window/42_trapping_rain_water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height: return 0
        water = 0
        n = len(height)
        l, r = 0, n - 1
        maxL, maxR = height[l], height[r]
        while l < r:
            if maxL < maxR:
                l += 1
                maxL = max(maxL, height[l])  
                water += maxL - height[l]
            else:
                r -= 1
                maxR = max(maxR, height[r])
                water += maxR - height[r]
        return water


        

File: monotonic_stack/42_trapping_rain_water.java
class Solution {
    public int trap(int[] height) {
        int res = 0;

        Stack<Integer> stack = new Stack<>();

        int curr = 0;
        while (curr < height.length) {
            while (!stack.empty() && height[curr] > height[stack.peek()]) {
                int topIdx = stack.pop();
                if (stack.empty()) break;

                int distance = curr - stack.peek() - 1;
                int boundedHeight = Math.min(height[curr], height[stack.peek()]) - height[topIdx];
                res += distance * boundedHeight;
            }
            stack.push(curr++);
        }

        return res;       
    }
}